lib/syncthing/utils.go

```diff
@@ -153,3 +159,79 @@ func copyFile(src, dst string) error {
 func OpenDBBackend(path string, tuning config.Tuning) (backend.Backend, error) {
 	return backend.Open(path, backend.Tuning(tuning))
 }
+
+// Opens a database and attempts migrating the legacy database to the new database format.
+func OpenDatabase(path string, oldDBDir string, evLogger events.Logger) (newdb.DB, error) {
+	sql, err := sqlite.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	sdb := newdb.MetricsWrap(sql)
+
+	if be, err := backend.OpenLevelDBRO(oldDBDir); err == nil {
+		// We have not migrated. We should do that.
+		migrateDatabase(be, evLogger, sdb, oldDBDir)
+		miscDB := dbext.NewMiscDB(sdb)
+		_ = miscDB.PutTime("migrated-from-leveldb-at", time.Now())
+		_ = miscDB.PutString("migrated-from-leveldb-by", build.LongVersion)
+	}
+
+	return sdb, nil
+}
+
+func migrateDatabase(be backend.Backend, evLogger events.Logger, sdb newdb.DB, oldDBDir string) {
+	l.Infoln("Migrating database from LevelDB to SQLite; this can take quite a while...")
+
+	ll, err := db.NewLowlevel(be, evLogger)
+	if err != nil {
+		l.Warnln("Failed to migrate:", err)
+		os.Exit(1)
+	}
+
+	for _, folder := range ll.ListFolders() {
+		l.Infoln("Migrating folder", folder, "...")
+		var batch []protocol.FileInfo
+		fs, err := db.NewFileSet(folder, ll)
+		if err != nil {
+			l.Warnln("Failed to migrate FileInfos:", err)
+			os.Exit(1)
+		}
+		snap, err := fs.Snapshot()
+		if err != nil {
+			l.Warnln("Failed to migrate FileInfos:", err)
+			os.Exit(1)
+		}
+		snap.WithHaveSequence(0, func(f protocol.FileInfo) bool {
+			batch = append(batch, f)
+			if len(batch) == 1000 {
+				if err := sdb.Update(folder, protocol.LocalDeviceID, batch); err != nil {
+					l.Warnln("Failed to migrate FileInfos:", err)
+					os.Exit(1)
+				}
+				batch = batch[:0]
+			}
+			return true
+		})
+		if len(batch) > 0 {
+			if err := sdb.Update(folder, protocol.LocalDeviceID, batch); err != nil {
+				l.Warnln("Failed to migrate FileInfos:", err)
+				os.Exit(1)
```

Comment:
So, this is in some sense just temporary code, but anyway -- once moved out of main() I don't think it's reasonable to have os.Exit in here. It should really return a status some kind so we can exit in the right places only

Reply:
Yeah, agree. I committed a fix. On iOS I especially don't want `os.Exit` so I can have an opportunity to show the actual error to the user.